File: core/shared/src/main/scala/diplomacy/variant/standard/rule/mutable/MovementOrderWithResult.scala
package diplomacy.variant.standard.rule.mutable

import diplomacy.board.Power
import diplomacy.variant.standard.rule.{ MovementOrderWithResult => BaseMovementOrderWithResult }
import diplomacy.variant.standard.rule._

class MovementOrderWithResult[Power_ <: Power](val order: MovementOrder[Power_])
    extends BaseMovementOrderWithResult[Power_] {
  def result: Option[Result] = this.innerResult
  def result_=(result: Result): Unit = {
    order match {
      case Order.Hold(_) =>
        result match {
          case Result.Dislodged(_) | Result.Success => this.innerResult = Option(result)
          case _ =>
        }
      case Order.Move(_, _, _) =>
        result match {
          case Result.Dislodged(_) | Result.Success =>
            this.innerResult match {
              case Some(Result.Success) =>
              case _ => this.innerResult = Option(result)
            }
          case Result.Failed | Result.Bounced =>
            this.innerResult match {
              case Some(Result.Dislodged(_)) =>
              case _ => this.innerResult = Option(result)
            }
          case _ =>
        }
      case Order.Support(_, _) =>
        result match {
          case Result.Dislodged(_) => this.innerResult = Option(result)
          case Result.Failed | Result.Success =>
            this.innerResult match {
              case Some(Result.Dislodged(_)) | Some(Result.Cut) | Some(Result.NoCorrespondingOrder) =>
              case _ => this.innerResult = Option(result)
            }
          case Result.Cut =>
            this.innerResult match {
              case Some(Result.Dislodged(_)) =>
              case _ => this.innerResult = Option(result)
            }
          case Result.NoCorrespondingOrder => this.innerResult = Option(result)
          case _ =>
        }
      case Order.Convoy(_, _) =>
        result match {
          case Result.Dislodged(_) => this.innerResult = Option(result)
          case Result.Failed | Result.Success =>
            this.innerResult match {
              case Some(Result.Dislodged(_)) | Some(Result.NoCorrespondingOrder) =>
              case _ => this.innerResult = Option(result)
            }
          case Result.NoCorrespondingOrder => this.innerResult = Option(result)
          case _ =>
        }
    }
  }

  def toImmutable: immutable.MovementOrderWithResult[Power_] = {
    immutable.MovementOrderWithResult(this.order, this.innerResult)
  }

  private[this] var innerResult: Option[Result] = None
}

File: src/variant/standard/rule/movement-order-with-result.ts
import { Result } from "./result"
import { Order, Hold, Move, Support, Convoy } from "./order"

export class MovementOrderWithResult<Power> {
  private result: Result | null
  constructor (public order: Order<Power>) {
    this.result = null
  }
  getResult (): Result | null {
    return this.result
  }
  setResult (result: Result) {
    if (this.order instanceof Hold) {
      if (result === Result.Dislodged || result === Result.Success) {
        this.result = result
      }
    } else if (this.order instanceof Move) {
      if (result === Result.Dislodged || result === Result.Success) {
        if (this.result !== Result.Success) {
          this.result = result
        }
      } else if (result === Result.Failed || result === Result.Bounced) {
        if (this.result !== Result.Dislodged) {
          this.result = result
        }
      }
    } else if (this.order instanceof Support) {
      if (result === Result.Dislodged) {
        this.result = result
      } else if (result === Result.Failed || result === Result.Success) {
        if (
          (this.result !== Result.Dislodged) &&
          (this.result !== Result.Cut) &&
          (this.result !== Result.NoCorrespondingOrder)
        ) {
          this.result = result
        }
      } else if (result === Result.Cut) {
        if (this.result !== Result.Dislodged) {
          this.result = result
        }
      } else if (result === Result.NoCorrespondingOrder) {
        this.result = result
      }
    } else if (this.order instanceof Convoy) {
      if (result === Result.Dislodged) {
        this.result = result
      } else if (result === Result.Failed || result === Result.Success) {
        if (
          this.result !== Result.Dislodged && this.result !== Result.NoCorrespondingOrder
        ) {
          this.result = result
        }
      } else if (result === Result.NoCorrespondingOrder) {
        this.result = result
      }
    } else {
      throw { err: `Invalid order: ${this.order}` }
    }
  }
}
